gcstorage/gcstorage.go

```diff
@@ -0,0 +1,275 @@
+		defaultKeyExpiry: cmp.Or(cfg.DefaultKeyExpiry, DefaultTTL),
+		client:           client,
+		bucketHandle:     client.Bucket(cfg.Bucket),
+	}, nil
+}
+
+func (g *GCStorage[V]) Exists(ctx context.Context, key string) (bool, error) {
+	attr, err := g.bucketHandle.Object(g.keyPrefix + key).Attrs(ctx)
+	if err != nil {
+		if errors.Is(err, storage.ErrObjectNotExist) {
+			return false, nil
+		}
+		return false, fmt.Errorf("cachestore/gcstorage: get attrs returned error: %w", err)
+	}
+
+	if attr.Metadata["expires_at"] != "" {
+		expiresAt, err := time.Parse(time.RFC3339, attr.Metadata["expires_at"])
+		if err != nil {
+			return false, fmt.Errorf("cachestore/gcstorage: time parse returned error: %w", err)
+		}
+		if !expiresAt.IsZero() && expiresAt.Before(time.Now()) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (g *GCStorage[V]) Set(ctx context.Context, key string, value V) error {
+	return g.SetEx(ctx, key, value, g.defaultKeyExpiry)
+}
+
+func (g *GCStorage[V]) SetEx(ctx context.Context, key string, value V, ttl time.Duration) error {
+	if len(key) > cachestore.MaxKeyLength {
+		return fmt.Errorf("cachestore/gcstorage: %w", cachestore.ErrKeyLengthTooLong)
+	}
+	if len(key) == 0 {
+		return fmt.Errorf("cachestore/gcstorage: %w", cachestore.ErrInvalidKey)
+	}
+
+	var expiresAt time.Time
+	if ttl != 0 {
+		expiresAt = time.Now().Add(ttl)
+	}
+
+	data, err := serialize(cacheObject[V]{
+		Object:    value,
+		ExpiresAt: expiresAt,
+	})
+	if err != nil {
+		return err
+	}
+
+	obj := g.bucketHandle.Object(g.keyPrefix + key)
+	w := obj.NewWriter(ctx)
+	w.ObjectAttrs.ContentType = "application/json"
+	w.ObjectAttrs.Metadata = map[string]string{
+		"expires_at": expiresAt.Format(time.RFC3339),
+	}
+	if _, err := w.Write(data); err != nil {
+		_ = w.Close()
+		return fmt.Errorf("cachestore/gcstorage: write returned error: %w", err)
+	}
+	if err := w.Close(); err != nil {
+		return fmt.Errorf("cachestore/gcstorage: close returned error: %w", err)
+	}
+	return nil
+}
+
+func (g *GCStorage[V]) Get(ctx context.Context, key string) (V, bool, error) {
+	value, _, exists, err := g.GetEx(ctx, key)
+	return value, exists, err
+}
+
+func (g *GCStorage[V]) GetEx(ctx context.Context, key string) (V, *time.Duration, bool, error) {
+	obj := g.bucketHandle.Object(g.keyPrefix + key)
+	r, err := obj.NewReader(ctx)
+	if err != nil {
+		if errors.Is(err, storage.ErrObjectNotExist) {
+			return *new(V), nil, false, nil
+		}
+		return *new(V), nil, false, fmt.Errorf("cachestore/gcstorage: new reader error: %w", err)
+	}
+	defer r.Close()
+
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return *new(V), nil, false, fmt.Errorf("cachestore/gcstorage: read all error: %w", err)
+	}
+
+	value, err := deserialize[cacheObject[V]](data)
+	if err != nil {
+		return *new(V), nil, false, fmt.Errorf("cachestore/gcstorage: deserialize error: %w", err)
+	}
+
+	if !value.ExpiresAt.IsZero() && value.ExpiresAt.Before(time.Now()) {
+		return *new(V), nil, false, nil
```

Comment:
IMHO we can delete the object, when it is expired. So we can return right away here next time: https://github.com/goware/cachestore/pull/18/files#diff-29986a0877798860831c64e94199e6fa49dc607e23d39bb20d898cc891aa2e48R147
```suggestion
	if !value.ExpiresAt.IsZero() && value.ExpiresAt.Before(time.Now()) {
			// delete the object when it is expired
		_ = g.Delete(ctx, key)

		return *new(V), nil, false, nil
```

And i am just thinking.
Do we have some mechanism to delete old expired objects? Or the expired objects would be just stacking in bucket?

Reply:
I thought about it as well. However, this will make this method slow. In the worst case it will execute 2 api calls.  So I just thought it's better to leave it in the bucket.

If it comes to cleaning objects:
-  the bucket itself may have [Object Lifecycle Policy](https://cloud.google.com/storage/docs/lifecycle) that will be applied on the bucket level. 
- we may implement an object cleaner

Having an object cleaner in this library would require some process to go through all of the objects and delete them when they are absolute. I can implement naive cleaner obj cleaner. However, there will be no way to stop it.

Reply:
For the speed, we can resolve it with running it in goroutine like this.
```go
go g.Delete(context.Background(), key)
```

For the cleaning, we can leverage this with `daysSinceCustomTime`?
https://cloud.google.com/storage/docs/metadata#custom-time

So when we create object. We can do something like this:
```go
	obj := g.bucketHandle.Object(fmt.Sprintf(g.keyPrefix + key)
	w := obj.NewWriter(ctx)
	w.ObjectAttrs.ContentType = "application/json"
	w.ObjectAttrs.Metadata = map[string]string{
		"expires_at": expiresAt.Format(time.RFC3339),
	}

	if !expiresAt.IsZero() {
		w.ObjectAttrs.CustomTime = expiresAt
	}

...
```

And we can set the lifecycle as this on initialization of bucket?
```json
{
  "lifecycle": {
    "rule": [
      {
        "action": {
          "type": "Delete"
        },
        "condition": {
          "matchesSuffix": [
            ":expiration",
          ],
          "daysSinceCustomTime": 1 // or 0?
        }
      }
    ]
  }
}
```

https://cloud.google.com/storage/docs/lifecycle#conditions
`daysSinceCustomTime`
The daysSinceCustomTime condition is satisfied when the specified number of days have passed since the date and time specified in an object's [Custom-Time metadata field](https://cloud.google.com/storage/docs/metadata#custom-time). For example, if an object's Custom-Time is 2020-05-16T10:00:00Z and the daysSinceCustomTime condition is 10 days, then the condition is satisfied for the object on and after 2020/05/26 10:00 UTC.

daysSinceCustomTime is never satisfied for an object with no Custom-Time metadata set.

This should delete automatically the old objects, after one day of expiration? I believe. Or maybe we can pass: `"daysSinceCustomTime": 0` adn the deletion would be right away after expiration.
But this must be tested.

Reply:
I think we can rely on bucket expiry policy by default

but let users call cleanup method explicitly if they need it?

```go
storage := gcstorage.New(...)
go storage.RemoveExpiredFilesEvery(ctx, time.Minute)
```

Reply:
I will implement both.

Reply:
We can't let users do that:
```go
func New[V any](cfg *Config, opts ...cachestore.StoreOptions) (cachestore.Store[V], error) {
```
We return them `cachestore.Store[V]` which is interface: https://github.com/goware/cachestore/blob/master/cachestore.go#L16

But we can add this `RemoveExpiredFilesEvery(ctx, time.Duration)` to that interface and add to other clients err with not supported, then it is possible.

And i believe that bucket expiry policy is `Object Lifecycle Management` which i mentioned in comment above.
Did you meant just set high age for bucket?

Reply:
can't we return `*struct[V]` implementation, rather than the interface?

Reply:
I think that would break the API, but let's add it interface. Why not?
Or it can be added to config on `gcstorage`, and start the goroutine in background on initialization of `gcstorage` client.

Reply:
but @david-littlefarmer mentioned, that we could also set expiry metadata on individual files ... is that possible? That'd be probably better than having a helper like `go storage.RemoveExpiredFilesEvery(ctx, time.Minute)`.